File: Java-SpringBoot/BE-02-DB/java-17-springboot-3.5/src/main/java/com/flyrank/taskservice/controller/TaskController.java
package com.flyrank.taskservice.controller;

import com.flyrank.taskservice.dto.*;
import com.flyrank.taskservice.service.TaskService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.*;

import java.util.List;

/**
 * REST Controller exposing Task CRUD API endpoints.
 */
@RestController
@RequestMapping("/tasks")
@Tag(name = "tasks", description = "Task Management CRUD Endpoints")
public class TaskController {

    private final TaskService taskService;

    public TaskController(TaskService taskService) {
        this.taskService = taskService;
    }

    @GetMapping(produces = MediaType.APPLICATION_JSON_VALUE)
    @Operation(summary = "List all tasks with optional search/filtering")
    public List<TaskResponse> getTasks(
            @RequestParam(name = "search", required = false) String search,
            @RequestParam(name = "done", required = false) Boolean done
    ) {
        return taskService.getAllTasks(search, done);
    }

    @GetMapping(value = "/{task_id}", produces = MediaType.APPLICATION_JSON_VALUE)
    @Operation(summary = "Get task by ID")
    public TaskResponse getTask(@PathVariable("task_id") Long taskId) {
        return taskService.getTaskById(taskId);
    }

    @PostMapping(consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
    @ResponseStatus(HttpStatus.CREATED)
    @Operation(summary = "Create a new task")
    public TaskResponse createTask(@Valid @RequestBody TaskCreateRequest request) {
        return taskService.createTask(request);
    }

    @PutMapping(value = "/{task_id}", consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
    @Operation(summary = "Replace task by ID (Full Replacement)")
    public TaskResponse updateTask(
            @PathVariable("task_id") Long taskId,
            @Valid @RequestBody TaskReplaceRequest request
    ) {
        return taskService.replaceTask(taskId, request);
    }

    @PatchMapping(value = "/{task_id}", consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
    @Operation(summary = "Partially update task by ID")
    public TaskResponse patchTask(
            @PathVariable("task_id") Long taskId,
            @RequestBody TaskUpdateRequest request
    ) {
        return taskService.patchTask(taskId, request);
    }

    @DeleteMapping("/{task_id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    @Operation(summary = "Delete task by ID")
    public void deleteTask(@PathVariable("task_id") Long taskId) {
        taskService.deleteTask(taskId);
    }
}

File: Java-SpringBoot/BE-02-DB/kotlin-2.3-springboot-3.5/src/main/kotlin/com/flyrank/taskservice/controller/TaskController.kt
package com.flyrank.taskservice.controller

import com.flyrank.taskservice.dto.*
import com.flyrank.taskservice.service.TaskService
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.tags.Tag
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/tasks")
@Tag(name = "tasks", description = "Task Management CRUD Endpoints (Kotlin 2.3 & K2 Compiler)")
class TaskController(
    private val taskService: TaskService
) {

    @GetMapping(produces = [MediaType.APPLICATION_JSON_VALUE])
    @Operation(summary = "List all tasks with optional search/filtering")
    fun getTasks(
        @RequestParam(name = "search", required = false) search: String?,
        @RequestParam(name = "done", required = false) done: Boolean?
    ): List<TaskResponse> {
        return taskService.getAllTasks(search, done)
    }

    @GetMapping(value = ["/{task_id}"], produces = [MediaType.APPLICATION_JSON_VALUE])
    @Operation(summary = "Get task by ID")
    fun getTask(@PathVariable("task_id") taskId: Long): TaskResponse {
        return taskService.getTaskById(taskId)
    }

    @PostMapping(consumes = [MediaType.APPLICATION_JSON_VALUE], produces = [MediaType.APPLICATION_JSON_VALUE])
    @ResponseStatus(HttpStatus.CREATED)
    @Operation(summary = "Create a new task")
    fun createTask(@Valid @RequestBody request: TaskCreateRequest): TaskResponse {
        return taskService.createTask(request)
    }

    @PutMapping(value = ["/{task_id}"], consumes = [MediaType.APPLICATION_JSON_VALUE], produces = [MediaType.APPLICATION_JSON_VALUE])
    @Operation(summary = "Replace task by ID (Full Replacement)")
    fun updateTask(
        @PathVariable("task_id") taskId: Long,
        @Valid @RequestBody request: TaskReplaceRequest
    ): TaskResponse {
        return taskService.replaceTask(taskId, request)
    }

    @PatchMapping(value = ["/{task_id}"], consumes = [MediaType.APPLICATION_JSON_VALUE], produces = [MediaType.APPLICATION_JSON_VALUE])
    @Operation(summary = "Partially update task by ID")
    fun patchTask(
        @PathVariable("task_id") taskId: Long,
        @RequestBody request: TaskUpdateRequest
    ): TaskResponse {
        return taskService.patchTask(taskId, request)
    }

    @DeleteMapping("/{task_id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    @Operation(summary = "Delete task by ID")
    fun deleteTask(@PathVariable("task_id") taskId: Long) {
        taskService.deleteTask(taskId)
    }
}
